File: P313 Super Ugly Number/Solution.java
class Solution {
    public int nthSuperUglyNumber(int n, int[] primes) {
        int[] uglyNumbers = new int[n];
        uglyNumbers[0] = 1;
        int primeLength = primes.length;
        int[] primeIndexs = new int[primeLength];
        
        int[] nextMultipleOfPrimes = Arrays.copyOf(primes, primeLength);
    
        for(int i = 1; i < n; i++) {
            uglyNumbers[i] = getMinValue(nextMultipleOfPrimes);
            for(int j = 0; j < primeLength; j++) {
                if(uglyNumbers[i] == nextMultipleOfPrimes[j]) {
                    primeIndexs[j]++;
                    nextMultipleOfPrimes[j] = uglyNumbers[primeIndexs[j]] * primes[j];
                }
            }
        }
        return uglyNumbers[n -1];
    }
    
    private static int getMinValue(int[] array) {
        int minValue = array[0];
        for (int i = 1; i < array.length; i++) {
            if (array[i] < minValue) {
                minValue = array[i];
            }
        }
        return minValue;
    }
}

File: P313 Super Ugly Number/solution.py
class Solution:
    def nthSuperUglyNumber(self, n: int, primes: List[int]) -> int:
        ugly_numbers = [0] * n
        ugly_numbers[0] = 1
        prime_indexs = [0] * len(primes)
        
        # copying a list using slicing
        next_multiple_of_primes = primes[:]
        for i in range(1, n):
            ugly_numbers[i] = min(next_multiple_of_primes)
            for j in range(len(primes)):
                if ugly_numbers[i] == next_multiple_of_primes[j]:
                    prime_indexs[j] += 1
                    next_multiple_of_primes[j] = ugly_numbers[prime_indexs[j]] * primes[j]
        return ugly_numbers[-1]
